lib-js/spec-directory/virtual-directory.ts

```diff
@@ -0,0 +1,203 @@
+import fs from "fs"
+import path from "path"
+import { Readable } from "stream"
+import SpecDirectory, { SpecIteratee } from "./spec-directory"
+import { archiveFromStream, Directory as HrxDirectory } from "node-hrx"
+import { RunOptions } from "../options"
+import { toHrx } from "./hrx"
+import { withAsyncCleanup } from "./cleanup"
+
+function createFileCache(dir: HrxDirectory) {
+  const cache: Record<string, string> = {}
+  for (const itemName of dir) {
+    const subitem = dir.get(itemName)
+    if (subitem?.isFile()) {
+      cache[itemName] = subitem.body
+    }
+  }
+  return cache
+}
+
+function createSubdirCache(dir: HrxDirectory) {
+  const cache: Record<string, HrxDirectory> = {}
+  for (const itemName of dir) {
+    const subitem = dir.get(itemName)
+    if (subitem?.isDirectory()) {
+      cache[itemName] = subitem
+    }
+  }
+  return cache
+}
+
+export default class VirtualDirectory extends SpecDirectory {
+  path: string
+  basePath: string
+  private fileNames: string[]
+  private fileCache: Record<string, string>
+  private subdirNames: string[]
+  private subdirCache: Record<string, HrxDirectory>
+  private isArchiveRoot: boolean
+  private modified = false
+
+  constructor(
+    basePath: string,
+    hrxDir: HrxDirectory,
+    root?: SpecDirectory,
+    parentOpts?: RunOptions
+  ) {
+    super(root, parentOpts)
+    this.path = path.resolve(basePath, hrxDir.path)
+    this.basePath = basePath
+    this.fileNames = hrxDir.list().filter((item) => hrxDir.get(item)?.isFile())
+    this.fileCache = createFileCache(hrxDir)
+    this.subdirNames = hrxDir
+      .list()
+      .filter((item) => hrxDir.get(item)?.isDirectory())
+    this.subdirCache = createSubdirCache(hrxDir)
+    this.isArchiveRoot = hrxDir.path === ""
+  }
+
+  // Factories
+
+  // Unarchive the given .hrx file and turn it into a spec path
+  static async fromArchive(
+    hrxPath: string,
+    root?: SpecDirectory,
+    parentOpts?: RunOptions
+  ) {
+    const stream = fs.createReadStream(hrxPath, { encoding: "utf-8" })
+    const archive = await archiveFromStream(stream)
+    const { dir, name } = path.parse(hrxPath)
+    return new VirtualDirectory(
+      path.resolve(dir, name),
+      archive,
+      root,
+      parentOpts
+    )
+  }
+
+  // Get an HRX archive from a string
+  static async fromContents(contents: string, path = "/tmp") {
+    const stream = Readable.from(contents)
+    const archive = await archiveFromStream(stream)
+    // TODO where should the temp path be?
```

Comment:
I'd require the caller to pass in a path explicitly. As far as sass-spec is concerned, a path is an integral component of a directory.

Reply:
I'm only using this method for testing, on methods where the actual path doesn't matter as much. Should I just make that clear in the documentation?

Reply:
Yeah, that's fine.